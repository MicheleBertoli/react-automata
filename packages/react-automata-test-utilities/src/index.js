/* eslint-disable import/prefer-default-export */

export { default as testStateMachine } from './testStateMachine'
