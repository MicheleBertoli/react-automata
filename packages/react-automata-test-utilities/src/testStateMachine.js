import idx from 'idx'
import React from 'react'
import TestRenderer from 'react-test-renderer'
import { getShortestPaths } from 'xstate/lib/graph'
import { stringify } from 'react-automata-utilities'

const testStateMachine = (Component, options = {}) => {
  const paths = getShortestPaths(Component.machine, options.extendedState)

  Object.keys(paths).forEach(key => {
    const initialData = idx(options, _ => _.fixtures.initialData)
    const renderer = TestRenderer.create(
      <Component initialData={initialData} />
    )
    const instance = renderer.getInstance()

    paths[key].forEach(({ event, state }) => {
      const fixtures = idx(options, _ => _.fixtures[state][event])

      instance.handleTransition(event, fixtures)
    })

    const machineState =
      instance.state.machineState.toString() ||
      stringify(instance.state.machineState.value).join(',')

    expect(renderer.toJSON()).toMatchSnapshot(machineState)
  })
}

export default testStateMachine
