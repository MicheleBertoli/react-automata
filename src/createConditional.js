import React from 'react'
import PropTypes from 'prop-types'
import { getContextValue } from './utils'

export const createConditional = ({
  displayName,
  propTypes,
  shouldShow,
  shouldHide,
}) => {
  class Conditional extends React.Component {
    constructor(props, context) {
      super(props, context)

      const value = getContextValue(context, props.channel)

      this.state = {
        visible: shouldShow(props, value),
      }

      if (this.state.visible && props.onEnter) {
        props.onEnter()
      }
    }

    componentWillReceiveProps(nextProps, nextContext) {
      const value = getContextValue(nextContext, nextProps.channel)

      if (!this.state.visible && shouldShow(nextProps, value)) {
        this.setState({
          visible: true,
        })

        if (nextProps.onEnter) {
          nextProps.onEnter()
        }
      }

      if (this.state.visible && shouldHide(nextProps, value)) {
        this.setState({
          visible: false,
        })

        if (nextProps.onLeave) {
          nextProps.onLeave()
        }
      }
    }

    render() {
      if (typeof this.props.render === 'function') {
        return this.props.render(this.state.visible)
      }

      return this.state.visible ? this.props.children : null
    }
  }

  Conditional.displayName = displayName

  Conditional.contextTypes = {
    automata: PropTypes.object,
  }

  Conditional.propTypes = {
    ...propTypes,
    channel: PropTypes.string,
    children: PropTypes.node,
    render: PropTypes.func,
    onEnter: PropTypes.func,
    onLeave: PropTypes.func,
  }

  Conditional.defaultProps = {
    children: null,
  }

  return Conditional
}

export default createConditional
