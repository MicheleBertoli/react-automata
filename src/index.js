export { default as Action } from './Action'
export { default as State } from './State'
export { default as testStateMachine } from './testStateMachine'
export { default as withStateMachine } from './withStateMachine'
