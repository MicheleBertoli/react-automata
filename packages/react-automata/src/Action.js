import createConditional from './createConditional'

export default createConditional('Action', 'actions')
