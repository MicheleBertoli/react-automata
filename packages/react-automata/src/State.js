import createConditional from './createConditional'

export default createConditional('State', 'machineState')
