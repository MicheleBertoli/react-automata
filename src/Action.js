import PropTypes from 'prop-types'
import createConditional from './createConditional'

const displayName = 'Action'

const propTypes = {
  hide: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.string,
  ]),
  show: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.string,
  ]),
}

const matches = (value, actions) =>
  Array.isArray(value)
    ? actions.some(action => value.includes(action))
    : actions.includes(value)

export const shouldShow = (props, context) =>
  matches(props.show, context.actions)

export const shouldHide = (props, context) =>
  props.hide
    ? matches(props.hide, context.actions)
    : !matches(props.show, context.actions)

export default createConditional({
  displayName,
  propTypes: process.env.NODE_ENV !== 'production' ? propTypes : null,
  shouldShow,
  shouldHide,
})
