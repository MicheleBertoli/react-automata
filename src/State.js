import PropTypes from 'prop-types'
import minimatch from 'minimatch'
import createConditional from './createConditional'

const displayName = 'State'

const propTypes = {
  value: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.string,
  ]),
}

const matches = (value, machineState) => {
  const values = Array.isArray(value) ? value : [value]
  const states = Array.isArray(machineState) ? machineState : [machineState]

  return values.some(val => states.some(state => minimatch(state, val)))
}

export const shouldShow = (props, context) =>
  matches(props.value, context.machineState)

export const shouldHide = (props, context) =>
  !matches(props.value, context.machineState)

export default createConditional({
  displayName,
  propTypes: process.env.NODE_ENV !== 'production' ? propTypes : null,
  shouldShow,
  shouldHide,
})
