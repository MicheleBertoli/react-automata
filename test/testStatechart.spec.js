import React from 'react'
import { Action, State, testStatechart, withStatechart } from '../src'

describe('conditional', () => {
  const secondMachine = {
    initial: 'a',
    states: {
      a: {
        on: {
          SECOND_NEXT: 'b',
        },
        onEntry: 'enterBA',
      },
      b: {
        on: {
          SECOND_NEXT: 'a',
        },
        onEntry: 'enterBB',
      },
    },
  }

  const firstMachine = {
    initial: 'a',
    states: {
      a: {
        on: {
          FIRST_NEXT: 'b',
        },
        onEntry: 'enterA',
      },
      b: {
        on: {
          FIRST_NEXT: 'a',
        },
        onEntry: 'enterB',
        ...secondMachine,
      },
    },
  }

  test('action', () => {
    const App = () => (
      <div>
        <Action show="enterA" hide="enterB">
          a
        </Action>
        <Action show="enterBA">b.a</Action>
        <Action show="enterBB">b.b</Action>
      </div>
    )

    testStatechart({ statechart: firstMachine }, App)
  })

  test('state', () => {
    const App = () => (
      <div>
        <State value="a">a</State>
        <State value="b.a">b.a</State>
        <State value="b.b">b.b</State>
      </div>
    )

    testStatechart({ statechart: firstMachine }, App)
  })
})

test('parallel', () => {
  const wordMachine = {
    parallel: true,
    states: {
      bold: {
        initial: 'off',
        states: {
          on: {
            on: { TOGGLE_BOLD: 'off' },
          },
          off: {
            on: { TOGGLE_BOLD: 'on' },
          },
        },
      },
      underline: {
        initial: 'off',
        states: {
          on: {
            on: { TOGGLE_UNDERLINE: 'off' },
          },
          off: {
            on: { TOGGLE_UNDERLINE: 'on' },
          },
        },
      },
      italics: {
        initial: 'off',
        states: {
          on: {
            on: { TOGGLE_ITALICS: 'off' },
          },
          off: {
            on: { TOGGLE_ITALICS: 'on' },
          },
        },
      },
      list: {
        initial: 'none',
        states: {
          none: {
            on: { BULLETS: 'bullets', NUMBERS: 'numbers' },
          },
          bullets: {
            on: { NONE: 'none', NUMBERS: 'numbers' },
          },
          numbers: {
            on: { BULLETS: 'bullets', NONE: 'none' },
          },
        },
      },
    },
  }

  const App = () => (
    <div>
      <State value="bold.on">bold.on</State>
      <State value="bold.off">bold.off</State>
      <State value="underline.on">underline.on</State>
      <State value="underline.off">underline.off</State>
      <State value="italics.on">italics.on</State>
      <State value="italics.off">italics.off</State>
      <State value="list.none">list.none</State>
      <State value="list.bullets">list.bullets</State>
      <State value="list.numbers">list.numbers</State>
    </div>
  )

  testStatechart({ statechart: wordMachine }, App)
})

test('channels', () => {
  const inner = {
    initial: 'inner',
    states: {
      inner: {},
    },
  }
  const Inner = () => (
    <div>
      <State channel="inner" value="inner">
        inner
      </State>
      <State channel="outer" value="outer">
        outer
      </State>
    </div>
  )
  const InnerMachine = withStatechart(inner, { channel: 'inner' })(Inner)

  const outer = {
    initial: 'outer',
    states: {
      outer: {},
    },
  }
  const Outer = () => (
    <div>
      <State channel="outer" value="outer">
        outer
      </State>
      <InnerMachine />
    </div>
  )

  testStatechart({ statechart: outer, channel: 'outer' }, Outer)
})

describe('cond', () => {
  const statechart = {
    initial: 'a',
    states: {
      a: {
        on: {
          EVENT: {
            b: {
              cond: extState => extState.shouldPass,
            },
          },
        },
      },
      b: {},
    },
  }

  const Cond = () => (
    <React.Fragment>
      <State value="a">A</State>
      <State value="b">B</State>
    </React.Fragment>
  )

  test('pass', () => {
    const extendedState = {
      shouldPass: true,
    }
    const fixtures = {
      a: {
        EVENT: extendedState,
      },
    }

    testStatechart({ statechart, fixtures, extendedState }, Cond)
  })

  test('fail', () => {
    testStatechart({ statechart }, Cond)
  })
})
