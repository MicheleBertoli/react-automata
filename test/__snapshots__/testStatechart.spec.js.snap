// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action: a 1`] = `
<div>
  a
</div>
`;

exports[`action: b.a 1`] = `
<div>
  b.a
</div>
`;

exports[`action: b.b 1`] = `
<div>
  b.b
</div>
`;

exports[`channels: outer 1`] = `
<div>
  outer
  <div>
    inner
    outer
  </div>
</div>
`;

exports[`parallel: bold.off,underline.off,italics.off,list.bullets 1`] = `
<div>
  bold.off
  underline.off
  italics.off
  list.bullets
</div>
`;

exports[`parallel: bold.off,underline.off,italics.off,list.none 1`] = `
<div>
  bold.off
  underline.off
  italics.off
  list.none
</div>
`;

exports[`parallel: bold.off,underline.off,italics.off,list.numbers 1`] = `
<div>
  bold.off
  underline.off
  italics.off
  list.numbers
</div>
`;

exports[`parallel: bold.off,underline.off,italics.on,list.bullets 1`] = `
<div>
  bold.off
  underline.off
  italics.on
  list.bullets
</div>
`;

exports[`parallel: bold.off,underline.off,italics.on,list.none 1`] = `
<div>
  bold.off
  underline.off
  italics.on
  list.none
</div>
`;

exports[`parallel: bold.off,underline.off,italics.on,list.numbers 1`] = `
<div>
  bold.off
  underline.off
  italics.on
  list.numbers
</div>
`;

exports[`parallel: bold.off,underline.on,italics.off,list.bullets 1`] = `
<div>
  bold.off
  underline.on
  italics.off
  list.bullets
</div>
`;

exports[`parallel: bold.off,underline.on,italics.off,list.none 1`] = `
<div>
  bold.off
  underline.on
  italics.off
  list.none
</div>
`;

exports[`parallel: bold.off,underline.on,italics.off,list.numbers 1`] = `
<div>
  bold.off
  underline.on
  italics.off
  list.numbers
</div>
`;

exports[`parallel: bold.off,underline.on,italics.on,list.bullets 1`] = `
<div>
  bold.off
  underline.on
  italics.on
  list.bullets
</div>
`;

exports[`parallel: bold.off,underline.on,italics.on,list.none 1`] = `
<div>
  bold.off
  underline.on
  italics.on
  list.none
</div>
`;

exports[`parallel: bold.off,underline.on,italics.on,list.numbers 1`] = `
<div>
  bold.off
  underline.on
  italics.on
  list.numbers
</div>
`;

exports[`parallel: bold.on,underline.off,italics.off,list.bullets 1`] = `
<div>
  bold.on
  underline.off
  italics.off
  list.bullets
</div>
`;

exports[`parallel: bold.on,underline.off,italics.off,list.none 1`] = `
<div>
  bold.on
  underline.off
  italics.off
  list.none
</div>
`;

exports[`parallel: bold.on,underline.off,italics.off,list.numbers 1`] = `
<div>
  bold.on
  underline.off
  italics.off
  list.numbers
</div>
`;

exports[`parallel: bold.on,underline.off,italics.on,list.bullets 1`] = `
<div>
  bold.on
  underline.off
  italics.on
  list.bullets
</div>
`;

exports[`parallel: bold.on,underline.off,italics.on,list.none 1`] = `
<div>
  bold.on
  underline.off
  italics.on
  list.none
</div>
`;

exports[`parallel: bold.on,underline.off,italics.on,list.numbers 1`] = `
<div>
  bold.on
  underline.off
  italics.on
  list.numbers
</div>
`;

exports[`parallel: bold.on,underline.on,italics.off,list.bullets 1`] = `
<div>
  bold.on
  underline.on
  italics.off
  list.bullets
</div>
`;

exports[`parallel: bold.on,underline.on,italics.off,list.none 1`] = `
<div>
  bold.on
  underline.on
  italics.off
  list.none
</div>
`;

exports[`parallel: bold.on,underline.on,italics.off,list.numbers 1`] = `
<div>
  bold.on
  underline.on
  italics.off
  list.numbers
</div>
`;

exports[`parallel: bold.on,underline.on,italics.on,list.bullets 1`] = `
<div>
  bold.on
  underline.on
  italics.on
  list.bullets
</div>
`;

exports[`parallel: bold.on,underline.on,italics.on,list.none 1`] = `
<div>
  bold.on
  underline.on
  italics.on
  list.none
</div>
`;

exports[`parallel: bold.on,underline.on,italics.on,list.numbers 1`] = `
<div>
  bold.on
  underline.on
  italics.on
  list.numbers
</div>
`;

exports[`state: a 1`] = `
<div>
  a
</div>
`;

exports[`state: b.a 1`] = `
<div>
  b.a
</div>
`;

exports[`state: b.b 1`] = `
<div>
  b.b
</div>
`;
